group/group.go

```diff
@@ -0,0 +1,75 @@
+// Package group provides prime-order groups based on elliptic curves.
+package group
+
+import (
+	"crypto/elliptic"
+	"encoding"
+	"errors"
+	"io"
+
+	"github.com/cloudflare/circl/ecc/p384"
+)
+
+var (
+	// P256 is the group generated by P-256 elliptic curve.
+	P256 Group = wG{elliptic.P256()}
+	// P384 is the group generated by P-384 elliptic curve.
+	P384 Group = wG{p384.P384()}
+	// P521 is the group generated by P-521 elliptic curve.
+	P521 Group = wG{elliptic.P521()}
+)
+
+// Group represents a prime-order group based on elliptic curves.
+type Group interface {
+	NewElement() Element
+	NewScalar() Scalar
+	Identity() Element
+	Generator() Element
+	Order() Scalar
+	Hashes([]byte) Hasher
+	Random() Randomizer
+}
+
+// Element represents an abstract element of a prime-order group.
+type Element interface {
+	IsIdentity() bool
+	IsEqual(a Element) bool
+	Add(a, b Element) Element
+	Dbl(a Element) Element
+	Neg(a Element) Element
+	Mul(a Element, n Scalar) Element
+	MulGen(n Scalar) Element
+	encoding.BinaryMarshaler
+	encoding.BinaryUnmarshaler
+	MarshalBinaryCompress() ([]byte, error)
+}
+
+// Scalar represents an integer scalar.
+type Scalar interface {
+	IsEqual(x Scalar) bool
+	Add(x, y Scalar) Scalar
+	Sub(x, y Scalar) Scalar
+	Mul(x, y Scalar) Scalar
+	Neg(Scalar) Scalar
+	Inv(Scalar) Scalar
+	encoding.BinaryMarshaler
+	encoding.BinaryUnmarshaler
+}
+
+// Randomizer allows to get random group elements and scalars.
+type Randomizer interface {
+	RndElt(io.Reader) Element
```

Comment:
Can we please spell these out? `RandomElement` and `RandomScalar`? (Also, `Randomizer` seems like a misleading name. Could we move these back to the `Group` interface?)